src/libYARP_OS/include/yarp/os/impl/Terminal.h

```diff
@@ -0,0 +1,186 @@
+#include <yarp/os/Bottle.h>
+#include <yarp/os/Vocab.h>
+
+#include <cstdio>
+#include <cstring>
+
+namespace yarp {
+namespace os {
+namespace impl {
+
+#ifdef WITH_LIBEDIT
+    #include <editline/readline.h>
+    #include <vector>
+    static std::vector<std::string> commands;
+    static yarp::os::Port* rpcHelpPort = nullptr;
+    static bool commandListInitialized = false;
+    static char* szLine = (char*)nullptr;
+    static bool readlineEOF=false;
+
+    static char* dupstr(char* s)
+    {
+        char *r;
+        r = (char*) malloc ((strlen (s) + 1));
+        strcpy (r, s);
+        return (r);
+    }
+    /* Generator function for command completion.  STATE lets us know whether
+   to start from scratch; without any state (i.e. STATE == 0), then we
+   start at the top of the list. */
+    static char* command_generator (const char* text, int state)
+    {
+        static size_t list_index, len;
+        char *name;
+
+        /* if this is a new word to complete, initialize now.  this includes
+            saving the length of text for efficiency, and initializing the index
+            variable to 0. */
+        if (!state)
+            {
+            list_index = 0;
+            len = strlen (text);
+            }
+
+        if (!commandListInitialized)
+        {
+            commands.clear();
+            yarp::os::Bottle helpCommand, helpBottle;
+            helpCommand.addString("help");
+            bool helpOk=false;
+            if (rpcHelpPort)
+                helpOk = rpcHelpPort->write(helpCommand, helpBottle);
+            if (helpOk)
+            {
+                yarp::os::Bottle* cmdList = nullptr;
+                if (helpBottle.get(0).isVocab() && helpBottle.get(0).asVocab()==VOCAB4('m', 'a', 'n', 'y') )
+                {
+                    cmdList=helpBottle.get(1).asList();
+                }
+                else
+                    cmdList=helpBottle.get(0).asList();
+                if (cmdList && cmdList->get(0).asString() == "*** Available commands:")
+                {
+                    for (int i=1; i<cmdList->size(); ++i)
+                        commands.push_back(cmdList->get(i).asString());
+                }
+            }
+            commands.push_back(" ");
+            commandListInitialized=true;
+        }
+        while ((list_index<commands.size()) && (name = (char*)commands[list_index].c_str()))
+            {
+            list_index++;
+            if (strncmp (name, text, len) == 0)
+                return (dupstr(name));
+            }
+
+        /* if no names matched, then return null. */
+        return ((char *)nullptr);
+    }
+    /* Attempt to complete on the contents of TEXT.  START and END show the
+   region of TEXT that contains the word to complete.  We can use the
+   entire line in case we want to do some simple parsing.  Return the
+   array of matches, or nullptr if there aren't any. */
+    static char ** my_completion (const char* text, int start, int end)
```

Comment:
This method generates a -Wunused-function warning, is this used anywhere?

Reply:
Actually `dupstr()` is only used by `generate_commands()`, which is only used by `my_completion()`. I see it is assigned here https://github.com/robotology/yarp/blob/9c8cea6fa292864ff2ee4aacf3a98553836cdc8d/src/libYARP_OS/src/Companion.cpp#L2191 
when `WITH_LIBEDIT` is set, to a pointer to function of the Readline library (explained here:
http://www.delorie.com/gnu/docs/readline/rlman_47.html)